source/slang/slang-check-expr.cpp

```diff
@@ -1529,6 +1538,51 @@ namespace Slang
         return CheckInvokeExprWithCheckedOperands(expr);
     }
 
+    Expr* SemanticsExprVisitor::visitTryExpr(TryExpr* expr)
+    {
+        auto prevTryClauseType = expr->tryClauseType;
+        m_enclosingTryClauseType = expr->tryClauseType;
+        expr->base = CheckTerm(expr->base);
+        m_enclosingTryClauseType = prevTryClauseType;
+        expr->type = expr->base->type;
+        if (as<ErrorType>(expr->type))
+            return expr;
+        
+        auto parentFunc = this->m_parentFunc;
+        // TODO: check if the try clause is catched.
+        // For now we assume all `try`s are not catched (because we don't have catch yet).
+        if (!parentFunc)
+        {
+            getSink()->diagnose(expr, Diagnostics::uncachedTryCallInNonThrowFunc);
+            return expr;
+        }
+        if (parentFunc->errorType->equals(m_astBuilder->getVoidType()))
+        {
+            getSink()->diagnose(expr, Diagnostics::uncachedTryCallInNonThrowFunc);
+            return expr;
+        }
+
+        if (!as<InvokeExpr>(expr->base))
+        {
+            getSink()->diagnose(expr, Diagnostics::tryClauseMustApplyToInvokeExpr);
+            return expr;
+        }
+        auto base = as<InvokeExpr>(expr->base);
+        if (auto callee = as<DeclRefExpr>(base->functionExpr))
+        {
+            if (auto funcCallee = as<FuncDecl>(callee->declRef.getDecl()))
+            {
+                if (funcCallee->errorType->equals(m_astBuilder->getVoidType()))
+                {
+                    getSink()->diagnose(expr, Diagnostics::tryInvokeCalleeShouldThrow, callee->declRef);
+                }
+                return expr;
```

Comment:
There seems to be a missing check here to ensure that the `errorType` of the callee is compatible with the `errorType` of the caller (unless the error is caught in the current scope).

For now I suppose we'd want a type equality test here (both error types must match exactly), but eventually it would be nice to have a type conversion check (although we'd need to figure out where in the AST to stash the conversion operation(s)).